core/src/style/textStyleBuilder.cpp

```diff
@@ -210,59 +208,240 @@ bool TextStyleBuilder::addFeature(const Feature& _feat, const DrawRule& _rule) {
+    // length of line needed for the label
+    const float labelLength = m_attributes.width * pixelSize;
+    // Allow labels to appear later than tile's min-zoom
+    const float minLength = labelLength * 0.6;
+
+    // Chord length for minimal ~120 degree inner angles (squared)
+    // sin(60)*2
+    const float sqDirLimit = powf(1.7f, 2);
+    // Range to check for angle changes
+    const float sampleWindow = pixelSize * 50;
+
+    // Minimal ~10 degree counts as change of direction
+    // cross(dir1,dir2) < sin(10)
+    const float flipTolerance = 0.17;
+
+    LineSampler<std::vector<glm::vec3>> sampler;
+
+    sampler.set(_line);
+
+    if (sampler.sumLength() < minLength) { return; }
+
+    struct LineRange {
+        size_t start, end;
+        int flips;
+        float area;
+    };
+
+    std::vector<LineRange> ranges;
+
+    for (size_t i = 0; i < _line.size()-1; i++) {
+
+        int flips = 0;
+        float lastAngle = 0;
+        float sumAngle = 0;
+        size_t lastBreak = 0;
+
+        glm::vec2 dir1 = sampler.segmentDirection(i);
 
-                    float d = sqPointSegmentDistance(pp, p1, p);
-                    if (d > tolerance) { break; }
+        for (size_t j = i + 1; j < _line.size()-1; j++) {
+            glm::vec2 dir2 = sampler.segmentDirection(j);
 
-                    // Skip merged segment
-                    merged = true;
-                    i += 1;
+            float angle = crossProduct(dir1, dir2);
+
+            if (std::abs(angle) > flipTolerance) {
+                if (lastAngle > 0) {
+                    if (angle < 0) { flips++; }
+                } else if (lastAngle < 0) {
+                    if (angle > 0) { flips++; }
                 }
-                p2 = p;
+                lastAngle = angle;
             }
 
-            // place labels at segment-subdivisions
-            int run = merged ? 1 : 2;
-            segmentLength /= run;
+            float length = sampler.point(j).z - sampler.point(i).z;
+
+            sumAngle += std::abs(angle);
 
-            while (segmentLength > minLength && run <= 4) {
-                glm::vec2 a = p1;
-                glm::vec2 b = glm::vec2(p2 - p1) / float(run);
+            // Limit number of direction changes (Avoid squiggly labels)
+            bool splitLine = flips > 2;
 
-                for (int r = 0; r < run; r++) {
-                    addLabel(_params, Label::Type::line, { a, a+b }, _rule);
-                    a += b;
+            if (!splitLine) {
+                // Go back within window to check for hard direction changes
+                for (int k = j - 1; k >= int(i); k--){
+                    if (glm::length2(sampler.segmentDirection(k) + dir2) < sqDirLimit) {
+                        //LOG("back break %s, %d %d", _params.text.c_str(), j, k);
+                        splitLine = true;
+                    }
+                    if (sampler.point(k).z < sampler.point(j).z - sampleWindow) {
```

Comment:
@hjanetzek can you explain the use of `sampleWindow` here? I am not sure I get it or maybe rename it to something which explains the context in which its used.